doc/api18970.md

```diff
@@ -0,0 +1,334 @@
+non-negative integer return value, but specific functions may use other
+conventions.
+
+An **operation** is behavior, which is generally initiated by invoking a
+function.  An operation can also succeed or fail.  An operation
+**completes** when the success or failure of the operation has been
+determined and made available through whatever mechanism was specified
+through the initiating function.
+
+For APIs we are particularly interested in the relationship between the
+function and its associated operation.  Some potential relationships
+include:
+* The function succeeds if and only if its associated operation
+  completes with success. (This is usually expected of *synchronous*
+  functions.)
+* If the function fails then its associated operation has not been
+  initiated.  If the function succeeds then its associated operation has
+  been initiated.  The success or failure of the operation is not
+  otherwise coupled to success or failure of its initiating function.
+  (This is usually expected of *asynchronous* functions.)
+
+## Definitions
+
+### rescheduling (function)
+
+A function is rescheduling when it invokes an operation that includes a
+reschedule point (potentially, but not necessarily, context switching to
+a non-current thread).
+
+Note that whether a reschedule-point function will cause a context
+switch depends on the priority of the current and ready threads.
+
+#### Commentary
+
+The term *rescheduling* applies only to functions that explicitly invoke
+operations that (may) encounter a reschedule point.  Unless interrupts
+are disabled any function may be context-switched if the interrupt
+causes a meta-irq thread to become ready.
+
+### reschedule-safe (function)
+
+A function is reschedule-safe if it is either not rescheduling, or if it
+is rescheduling but no reschedule point reached during its execution
+would cause a cooperative thread invoking it to be context-switched.
+
+#### Commentary
+
+For thread safety it is generally necessary to hold a lock of some sort.
+If a sequence of code involves only invocation of functions that do not
+context-switch then the lock may be avoidable.
+
+See also discussion of the term "block" at the definition of
+*synchronous*.
+
+Note that if a function is marked as isr-callable then it
+must internally check `k_is_in_isr()` and behave as a no-wait function,
+whether or not it was invoked in a no-wait mode.
+
+### context-switching (function)
+
+A function is context-switching when it is rescheduling and is
+guaranteed to cause a context switch when the head of the ready queue is
+not the current thread, even for cooperative threads.
+
+#### Commentary
+
+Examples include `k_yield()` and `k_sleep()` (TBD: only when passed a
+non-zero timeout?).
+
+### no-wait (function)
+
+A rescheduling function that is not reschedule-safe is no-wait when
+calls to it can be forced by a parameter into a reschedule-safe path,
+i.e. one that will not trigger a context switch from a cooperative
```

Comment:
For example, using "cooperative-safe" would probably make this clause unnecessary.

The above would read "a rescheduling function that is not cooperative-safe is no-wait when calls to it can be forced by a parameter into a cooperative-safe path," which I think is pretty clear on its own.

Reply:
Revised to use `cooperative-unsafe`.